[SwiftUI] Re-using the same WebPage for a different WebView after the original WebView has been destroyed causes a crash

rr-codes opened the pull request:
#### f6dd331abca3754d4fe18fa43cd33ee38c88fb6d
<pre>
[SwiftUI] Re-using the same WebPage for a different WebView after the original WebView has been destroyed causes a crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=289561">https://bugs.webkit.org/show_bug.cgi?id=289561</a>
<a href="https://rdar.apple.com/146719280">rdar://146719280</a>

Reviewed by Tim Horton and Abrar Rahman Protyasha.

While it is invalid to have multiple visible WebViews using the same WebPage, it is perfectly valid for them to use a WebPage one at a time.

To fix this, implement the `dismantle{UI/NS}View` protocol method and inform the WebPage that it is no longer bounded to a WebView.

* Source/WebKit/_WebKit_SwiftUI/Implementation/WebViewRepresentable.swift:
(WebViewRepresentable.dismantlePlatformView(_:coordinator:)):
(WebViewRepresentable.dismantleUIView(_:coordinator:)):
(WebViewRepresentable.dismantleNSView(_:coordinator:)):

Canonical link: <a href="https://commits.webkit.org/291994@main">https://commits.webkit.org/291994@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f93bddda321ea7e984cd2aa38791547870f56ddd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94588 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14180 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3981 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99609 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45103 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14477 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22608 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72170 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29484 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97590 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10776 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85430 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52501 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10470 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44427 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80686 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3207 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101651 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21642 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15788 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81172 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21891 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81422 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80548 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25106 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2495 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14859 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15192 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21620 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26744 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21294 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24759 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23032 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

axiixc reviewed (commented):
Is there a separate task tracking diagnostics for invalid reuse as well? If this crashes I’d expect those do too. Ex…
```swift
HSplitView { WebView(page); WebView(page); }
```

axiixc commented on Source/WebKit/_WebKit_SwiftUI/Implementation/WebViewRepresentable.swift:
```diff
@@ -113,6 +113,10 @@ struct WebViewRepresentable {
     }
+
+    static func dismantlePlatformView(_ platformView: CocoaWebViewAdapter, coordinator: WebViewCoordinator) {
+        coordinator.configuration.owner.page.isBoundToWebView = false
```
Maybe something for the future: I wonder if instead of just a boolean we could use some sort of `@Namespace` value inside the representable which would let us define both owned/unowned while also allowing for us to verify _expected_ ownership via `==`.

rr-codes replied:
That's a good idea!

rr-codes committed:
[SwiftUI] Re-using the same WebPage for a different WebView after the original WebView has been destroyed causes a crash
https://bugs.webkit.org/show_bug.cgi?id=289561
rdar://146719280

Reviewed by Tim Horton and Abrar Rahman Protyasha.

While it is invalid to have multiple visible WebViews using the same WebPage, it is perfectly valid for them to use a WebPage one at a time.

To fix this, implement the `dismantle{UI/NS}View` protocol method and inform the WebPage that it is no longer bounded to a WebView.

* Source/WebKit/_WebKit_SwiftUI/Implementation/WebViewRepresentable.swift:
(WebViewRepresentable.dismantlePlatformView(_:coordinator:)):
(WebViewRepresentable.dismantleUIView(_:coordinator:)):
(WebViewRepresentable.dismantleNSView(_:coordinator:)):

Canonical link: https://commits.webkit.org/291994@main
